mqtt/mqtt-broker/src/sys_topic.rs

```diff
@@ -0,0 +1,389 @@
+                Session::Persistent(_) => Some(client_id),
+                _ => None,
+            })
+            .collect();
+
+        Self::Connections(connections)
+    }
+
+    pub fn new_session(sessions: &'a HashMap<ClientId, Session>) -> Self {
+        let sessions = sessions
+            .iter()
+            .filter_map(|(client_id, session)| match session {
+                Session::Transient(_) => Some(client_id),
+                Session::Persistent(_) => Some(client_id),
+                Session::Offline(_) => Some(client_id),
+                _ => None,
+            })
+            .collect();
+
+        Self::Sessions(sessions)
+    }
+}
+
+impl<'a> TryFrom<StateChange<'a>> for proto::Publication {
+    type Error = Error;
+
+    fn try_from(state: StateChange<'a>) -> Result<Self, Error> {
+        Ok(match state {
+            StateChange::Subscriptions(client_id, subscriptions) => {
+                let payload = if let Some(subscriptions) = subscriptions {
+                    serde_json::to_string(&subscriptions)
+                        .map_err(|_| Error::NotifyError)?
+                        .into()
+                } else {
+                    "".into()
+                };
+
+                proto::Publication {
+                    topic_name: format!("$edgehub/subscriptions/{}", client_id),
+                    qos: STATE_CHANGE_QOS,
+                    retain: true,
+                    payload,
+                }
+            }
+            StateChange::Connections(connections) => proto::Publication {
+                topic_name: "$edgehub/connected".to_owned(),
+                qos: STATE_CHANGE_QOS,
+                retain: true,
+                payload: serde_json::to_string(&connections)
+                    .map_err(|_| Error::NotifyError)?
+                    .into(),
+            },
+            StateChange::Sessions(sessions) => proto::Publication {
+                topic_name: "$edgehub/sessions".to_owned(),
+                qos: STATE_CHANGE_QOS,
+                retain: true,
+                payload: serde_json::to_string(&sessions)
+                    .map_err(|_| Error::NotifyError)?
+                    .into(),
+            },
+        })
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::*;
+    use crate::broker::tests::*;
+    use crate::session::*;
+    use crate::subscription::*;
+    use crate::*;
+    use std::collections::{HashMap, VecDeque};
+    use std::convert::TryInto;
+    use std::iter::FromIterator;
+    use std::str::FromStr;
+    #[test]
+    fn test_subscriptions() {
+        let expected_id: ClientId = "Session".into();
+
+        // Session with no subscriptions
+        let session = make_session(expected_id.as_str(), Vec::<&str>::new(), true);
+        let no_subs = StateChange::new_subscription(&expected_id, &session);
+
+        if let StateChange::Subscriptions(stored_id, stored_subs) = &no_subs {
```

Comment:
isn't compiler smart enough to borrow here automatically?
```suggestion
        if let StateChange::Subscriptions(stored_id, stored_subs) = no_subs {
```

Reply:
```
error[E0382]: use of moved value: `many_subs`                                                     
   --> mqtt-broker\src\sys_topic.rs:154:43
    |
147 |         if let StateChange::Subscriptions(stored_id, stored_subs) = many_subs {
    |                                                      ----------- value moved here
...
154 |         let message: proto::Publication = many_subs.try_into().unwrap();
    |                                           ^^^^^^^^^ value used here after partial move      
    |
    = note: move occurs because value has type `std::option::Option<std::vec::Vec<&str>>`, which does not implement the `Copy` trait
```